### ISSUE TITLE
unexpected behavior of sudo condition

### ISSUE THREAD

**Issue by umputun** [umputun/spot#245]
### Discussed in https://github.com/umputun/spot/discussions/244

<div type='discussions-op-text'>

<sup>Originally posted by **dstull** December 20, 2024</sup>
@umputun I'm currently trying to compact some of my items down into a smaller list of tasks. To do that, I need to run some commands in the `cond` area that require `sudo` to be set.

- current config:
```yml
user: ubuntu
ssh_key: /home/ubuntu/.ssh/id_rsa
ssh_shell: /bin/bash
local_shell: /bin/bash
inventory: inventory.yml

tasks:
  - name: Check known rails issue 366450
    targets: ["gitlab_rails"]
    commands:
      - name: Condition check
        script: |
          rm -f /tmp/366450
          version=${GITLAB_VERSION//./}
          if [[ $version -gt 1500 && $version -lt 1630 ]]; then
            touch /tmp/366450
          fi
      - name: Checking for known issue
        script: |
          if grep -s 'OmniAuth::Strategies::SAML' /etc/gitlab/gitlab.rb | grep strategy_class; then
            touch /tmp/366450_warning
          fi
        on_exit: "rm -fv /tmp/366450"
        options: {ignore_errors: true, sudo: true}
        cond: "test -f /tmp/366450"
      - name: Display warning message
        echo: "If you're using SAML for OmniAuth you might be running into known issue https://gitlab.com/gitlab-org/gitlab/-/issues/366450"
        cond: "test -f /tmp/366450_warning"
      - name: Cleanup temp files
        script: |
          rm -fv /tmp/366450_warning
        options: {ignore_errors: true, sudo: true}
  - name: Fake issue that will fail
    targets: ["gitlab_rails"]
    commands:
      - name: Condition check
        script: |
          rm -f /tmp/fake
          version=${GITLAB_VERSION//./}
          if [[ $version -gt 1500 && $version -lt 1700 ]]; then
            touch /tmp/fake
          fi
      - name: Checking condition for known fake issue
        script: |
          if grep -s "gitlab_backup_cli\['dir'\]" /etc/gitlab/gitlab.rb | grep '/var/opt/gitlab/backups'; then
            touch /tmp/fake_warning
          fi
        on_exit: "rm -fv /tmp/fake"
        options: {ignore_errors: true, sudo: true}
        cond: "test -f /tmp/fake"
      - name: Display warning message
        echo: "If this is set to backups, that is wrong. Make sure you fix that."
        cond: "test -f /tmp/fake_warning"
      - name: Cleanup temp files
        script: |
          rm -fv /tmp/fake_warning
        options: {ignore_errors: true, sudo: true}
  - name: Fake issue that will fail with enhanced cond use
    targets: ["gitlab_rails"]
    commands:
      - name: Checking condition for known fake issue
        script: |
          if grep -s "gitlab_backup_cli\['dir'\]" /etc/gitlab/gitlab.rb | grep '/var/opt/gitlab/backups'; then
            touch /tmp/fake_warning
          fi
        options: {ignore_errors: true, sudo: true}
        cond: |
          version=${GITLAB_VERSION//./}
          [[ $version -gt 1500 && $version -lt 1700 ]]
      - name: Display warning message
        echo: "If this is set to backups, that is wrong. Make sure you fix that."
        cond: "test -f /tmp/fake_warning"
      - name: Cleanup temp files
        script: |
          rm -fv /tmp/fake_warning
        options: {ignore_errors: true, sudo: true}
  - name: Fake issue that will fail with compacted commands style
    targets: ["gitlab_rails"]
    commands:
      - name: Display warning message
        echo: "If this is set to backups, that is wrong. Make sure you fix that."
        options: { ignore_errors: true, sudo: true }
        cond: |
          version=${GITLAB_VERSION//./}
          [[ $version -gt 1500 && $version -lt 1700 && grep -s "gitlab_backup_cli\['dir'\]" /etc/gitlab/gitlab.rb | grep '/var/opt/gitlab/backups' ]]
```

- output. This should not skip the last command as the `grep` does get a value if `sudo` is used :

```console
ubuntu@sr-env-049aca04-omnibus:~$ ./spot -p rails_known_issues.yml --env GITLAB_VERSION:16.0.0
spot v1.16.2-1cf9c8b-2024-12-19T21:34:18Z
[omnibus localhost:22] run task "Check known rails issue 366450", commands: 4
[omnibus localhost:22] completed command "Condition check" {script: /bin/sh -c [multiline script]} (717ms)
[omnibus localhost:22] completed command "Checking for known issue" {script: /bin/sh -c [multiline script]} (335ms)
[omnibus localhost:22] completed command "Display warning message" {skip: Display warning message} (5ms)
[omnibus localhost:22] completed command "Cleanup temp files" {script: /bin/sh -c [multiline script]} (36ms)
[omnibus localhost:22] completed task "Check known rails issue 366450", commands: 4 (1.147s)
[omnibus localhost:22] run task "Fake issue that will fail", commands: 4
[omnibus localhost:22] completed command "Condition check" {script: /bin/sh -c [multiline script]} (226ms)
[omnibus localhost:22] completed command "Checking condition for known fake issue" {script: /bin/sh -c [multiline script]} (53ms)
[omnibus localhost:22] completed command "Display warning message" {echo: If this is set to backups, that is wrong. Make sure you fix that.} (9ms)
[omnibus localhost:22] completed command "Cleanup temp files" {script: /bin/sh -c [multiline script]} (35ms)
[omnibus localhost:22] completed task "Fake issue that will fail", commands: 4 (382ms)
[omnibus localhost:22] run task "Fake issue that will fail with enhanced cond use", commands: 3
[omnibus localhost:22] completed command "Checking condition for known fake issue" {script: /bin/sh -c [multiline script]} (290ms)
[omnibus localhost:22] completed command "Display warning message" {echo: If this is set to backups, that is wrong. Make sure you fix that.} (9ms)
[omnibus localhost:22] completed command "Cleanup temp files" {script: /bin/sh -c [multiline script]} (35ms)
[omnibus localhost:22] completed task "Fake issue that will fail with enhanced cond use", commands: 3 (395ms)
[omnibus localhost:22] run task "Fake issue that will fail with compacted commands style", commands: 1
[omnibus localhost:22] completed command "Display warning message" {skip: Display warning message} (270ms)
[omnibus localhost:22] completed task "Fake issue that will fail with compacted commands style", commands: 1 (328ms)
```

Further, If I remove the `grep` conditional like this:

```yml
  - name: Fake issue that will fail with compacted commands style
    targets: ["gitlab_rails"]
    commands:
      - name: Display warning message
        echo: "If this is set to backups, that is wrong. Make sure you fix that."
        options: { ignore_errors: true, sudo: true }
        cond: |
          version=${GITLAB_VERSION//./}
          [[ $version -gt 1500 && $version -lt 1700 ]]
```

I see that `sudo` use in general for an `echo` command seems to throw an error too, which seems surprising:

```console
[omnibus localhost:22] run task "Fake issue that will fail with compacted commands style", commands: 1
[omnibus localhost:22] run command "Display warning message"
[omnibus localhost:22]  > sudo /bin/sh -c "/bin/sh -c /tmp/.spot-4197468746377705472/spot-script3140008356"
[omnibus localhost:22]  > sudo "echo \"If this is set to backups, that is wrong. Make sure you fix that.\""
[omnibus localhost:22]  ! sudo: echo "If this is set to backups, that is wrong. Make sure you fix that.": command not found
[omnibus localhost:22] failed command "Display warning message" (241ms)
[omnibus localhost:22] completed task "Fake issue that will fail with compacted commands style", commands: 0 (301ms)
```

So then I remove the `sudo` option just to test that theory out and it runs fine, confirming that `sudo` on an `echo` doesn't seem to be supported:

```yml
  - name: Fake issue that will fail with compacted commands style
    targets: ["gitlab_rails"]
    commands:
      - name: Display warning message
        echo: "If this is set to backups, that is wrong. Make sure you fix that."
        options: { ignore_errors: true }
        cond: |
          version=${GITLAB_VERSION//./}
          [[ $version -gt 1500 && $version -lt 1700 ]]
```

```console
[omnibus localhost:22] run task "Fake issue that will fail with compacted commands style", commands: 1
[omnibus localhost:22] completed command "Display warning message" {echo: If this is set to backups, that is wrong. Make sure you fix that.} (224ms)
[omnibus localhost:22] completed task "Fake issue that will fail with compacted commands style", commands: 1 (281ms)
```

So then in the end my questions are:
1. Do you think `sudo` could be applied to `cond` run?
2. Do you think my use case makes sense enough to think `echo` should be able to support running correctly with `sudo` option set?

</div>

**Comment by umputun** [comment]
https://github.com/umputun/spot/issues/245

**Comment by umputun** [comment]
@dstull  this is what I was able to figure out so far:

- `echo` with sudo is supposed to work, but I have never tried to use it this way and see it as something very exotic. The fact that echo with sudo failed with "command not found" was a bug in a part of the code applying sudo to the final echo command, and it has been fixed. Generally speaking, having such support for echo makes sense, at least for consistency. Your use case, where echo itself doesn't actually need sudo but the condition requires it, is definitely a valid one. Another one I can think of is to do something crazy like this `echo $(/etc/shadow)` with sudo.

- Condition check with sudo is less clear to me. I don't see any obvious issues here, and it has good tests verifying this behavior. I just added one more like this, and it worked (condition passed) with the sudo option and failed (condition caused skip) without, as expected.

```go
	t.Run("condition true, sudo only", func(t *testing.T) {
		// check without sudo condition on root-only file
		ec := execCmd{exec: sess, tsk: &config.Task{Name: "test"}, cmd: config.Cmd{Condition: "cat /etc/shadow",
			Script: "echo condition true", Name: "test"}}
		resp, err := ec.Script(ctx)
		require.NoError(t, err)
		assert.Equal(t, " {skip: test}", resp.details)

		// check with sudo condition on root-only file
		ec = execCmd{exec: sess, tsk: &config.Task{Name: "test"}, cmd: config.Cmd{Condition: "cat /etc/shadow",
			Script: "echo condition true", Name: "test", Options: config.CmdOptions{Sudo: true}}}
		resp, err = ec.Script(ctx)
		require.NoError(t, err)
		assert.Equal(t, " {script: /bin/sh -c 'echo condition true', sudo: true}", resp.details)
	})
	```

From your large example, it is a little hard to extract a minimal part that may help me reproduce the issue. Can you isolate the task with a single command that I will be able to run to understand what's going on? From your output without debug and verbosity, I don't have much to work with.

**Comment by dstull** [comment]
@umputun thanks for handling the `echo` fix!

> Condition check with sudo

I believe that was an error in my command conditional setup and there are no issues with spot

Here is how I tested to prove out:
1. setup a file that only `root` could read with `this` as contents

```console
ubuntu@sr-env-049aca04-omnibus:~$ ls -ltr /tmp/root_only_file.txt
-rw------- 1 root root 14 Dec 20 21:54 /tmp/root_only_file.txt
```

2. simplified config(not using `echo` since it needs a fix as you state above, so I use `script` for now)

```yml
user: ubuntu
ssh_key: /home/ubuntu/.ssh/id_rsa
ssh_shell: /bin/bash
local_shell: /bin/bash
inventory: inventory.yml

tasks:
  - name: Desired ultimate compacted style that does not work right now with sudo - https://github.com/umputun/spot/discussions/244
    targets: ["gitlab_rails"]
    commands:
      - name: Display warning message
        echo: "If this is set to backups, that is wrong. Make sure you fix that."
        options: { ignore_errors: true, sudo: true }
        cond: |
          version=${GITLAB_VERSION//./}
          [[ $version -gt 1500 && $version -lt 1700 && grep -s this /tmp/root_only_file.txt ]]
```

3. Ran that and noticed it skipping when it should have not been skipped

```console
ubuntu@sr-env-049aca04-omnibus:~$ ./spot -p example.yml --env GITLAB_VERSION:16.0.0
spot v1.16.2-1cf9c8b-2024-12-19T21:34:18Z
[omnibus localhost:22] run task "Desired ultimate compacted style that does not work right now with sudo - https://github.com/umputun/spot/discussions/244", commands: 1
[omnibus localhost:22] completed command "Display warning message" {skip: Display warning message} (821ms)
[omnibus localhost:22] completed task "Desired ultimate compacted style that does not work right now with sudo - https://github.com/umputun/spot/discussions/244", commands: 1 (874ms)
```

At that point I remembered some other issues I had, scripting wise, and realized the command needed `` around it due to spaces.

Changing it to:

```yml
user: ubuntu
ssh_key: /home/ubuntu/.ssh/id_rsa
ssh_shell: /bin/bash
local_shell: /bin/bash
inventory: inventory.yml

tasks:
  - name: Desired ultimate compacted style that does not work right now with sudo - https://github.com/umputun/spot/discussions/244
    targets: ["gitlab_rails"]
    commands:
      - name: Display warning message
        script: echo "If this is set to backups, that is wrong. Make sure you fix that."
        options: { ignore_errors: true, sudo: true }
        cond: |
          version=${GITLAB_VERSION//./}
          [[ $version -gt 1500 && $version -lt 1700 && `grep -s this /tmp/root_only_file.txt` ]]
```

allows it to work as expected

```console
ubuntu@sr-env-049aca04-omnibus:~$ ./spot -p example.yml --env GITLAB_VERSION:16.0.0
spot v1.16.2-1cf9c8b-2024-12-19T21:34:18Z
[omnibus localhost:22] run task "Desired ultimate compacted style that does not work right now with sudo - https://github.com/umputun/spot/discussions/244", commands: 1
[omnibus localhost:22] completed command "Display warning message" {script: /bin/bash -c 'export GITLAB_VERSION="16.0.0"; echo "If this is set to backups, that is wrong. Make sure you fix that."', sudo: true} (434ms)
[omnibus localhost:22] completed task "Desired ultimate compacted style that does not work right now with sudo - https://github.com/umputun/spot/discussions/244", commands: 1 (487ms)
```

**Comment by umputun** [comment]
Still a bit confused. The issue (if you still have one) doesn't seem to be related to the spot itself but rather the condition expression. I have made this simple test case with your example:

```
  - name: sudo-cond
    targets: ["dev1"]
    commands:
      - name: Display warning message
        echo: "If this is set to backups, that is wrong. Make sure you fix that."
        options: { ignore_errors: true, sudo: true }
        cond: |
          version=${GITLAB_VERSION//./}
          echo "version is $version"
          echo "is is $(id)"
          [[ $version -gt 1500 && $version -lt 1700 && `grep -s this /tmp/root_only_file.txt` ]]
```

running it with -vv and --dbg gives me this (kept the relevalt part only):

```
[[dev1 dev1.umputun.dev:22]  > version is 1600
[dev1 dev1.umputun.dev:22]  > id is uid=0(root) gid=0(root) groups=0(root),1(bin),2(daemon),3(sys),4(adm),6(disk),10(wheel),11(floppy),20(dialout),26(tape),27(video)
2024/12/20 17:16:04.620 [DEBUG] removed local temp script /var/folders/fx/rzs1_n8137qfktxcbt_2v8pc0000gp/T/spot-script2293970068
2024/12/20 17:16:04.657 [INFO]  deleted recursively /tmp/.spot-6766569064362345472 from dev1
2024/12/20 17:16:04.659 [DEBUG] run sudo /bin/sh -c 'echo "If this is set to backups, that is wrong. Make sure you fix that."'
2024/12/20 17:16:04.659 [DEBUG] run ssh command "sudo /bin/sh -c 'echo \"If this is set to backups, that is wrong. Make sure you fix that.\"'" on 192.168.1.41:22
[dev1 dev1.umputun.dev:22]  > sudo /bin/sh -c 'echo "If this is set to backups, that is wrong. Make sure you fix that."'
[dev1 dev1.umputun.dev:22]  > If this is set to backups, that is wrong. Make sure you fix that.
[dev1 dev1.umputun.dev:22] completed command "Display warning message" {echo: If this is set to backups, that is wrong. Make sure you fix that.} (98ms)
[dev1 dev1.umputun.dev:22] completed task "sudo-cond", commands: 1 (178ms)
```

So, nothing wrong here, at least it looks sane to me. Running you prev version (without `` surouding grep part) showed bash error ` ! /tmp/.spot-4437783034070509056/spot-script2965106840: line 7: conditional binary operator expected` 

You can test all of this manually by making the same script spot is making, i.e.

```sh
#!/bin/bash
set -e
export GITLAB_VERSION="16.0.0"
version=${GITLAB_VERSION//./}
[[ $version -gt 1500 && $version -lt 1700 && `grep -s this /tmp/root_only_file.txt` ]]
```

btw, I'm not sure if the use of bashisms in the condition is intentional (it probably is, as you set this shell explicitly), but the same result can be achieved with POSIX sh-compatible syntax, like this:

```yml
        cond: |
          version=$(echo "$GITLAB_VERSION" | tr -d '.')
          echo "version is $version"
          echo "is is $(id)"
          [ "$version" -gt 1500 ] && [ "$version" -lt 1700 ] && grep -s this /tmp/root_only_file.txt          
```

**Comment by umputun** [comment]
probably i missed this line: "I believe that was an error in my command conditional setup and there are no issues with spot"

it has been merged with echo sudo fix, v1.6.3 released

**Comment by dstull** [comment]
Thanks so much!
